### PR TITLE
Added avatar editor

### LINKED ISSUE #1669: Create FC for the avatar window
- сreate `FC` for the avatar window
- display better `tooltip`s

### PR DESCRIPTION
**Pull request by Cheshiriks**
- added avatar editor

Closes #1669

### CODE REVIEW

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt [hunk 1]
```diff
@@ -474,9 +479,9 @@ class OrganizationView : AbstractView<OrganizationProps, OrganizationViewState>(
             it.decodeFromJsonString()
         }
 
-    private fun postImageUpload(element: HTMLInputElement) =
+    private fun postImageUpload(file: File) =
             scope.launch {
-                element.files!!.asList().single().let { file ->
+                file.let { file ->
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt:484]: `.let` can be omitted

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt [hunk 2]
```diff
@@ -507,6 +511,32 @@ class OrganizationView : AbstractView<OrganizationProps, OrganizationViewState>(
...
+        modalAvatarBuilder(
+            isOpen = state.isAvatarWindowOpen,
+            title = "Change organization's avatar",
+            onCloseButtonPressed = {
+                setState {
+                    isAvatarWindowOpen = false
+                }
+            }
+        ) {
+            div {
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt:523]: this part looks a common for `OrganizationView` and `UserSettingsView` -- maybe move it to modalAvatarBuilder?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt (hunk 1 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt]: Removed it

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt]: Yes. Moved it

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt [hunk 3]
```diff
@@ -271,3 +256,69 @@ fun ChildrenBuilder.modalBuilder(
         }
     }
 }
+
+/**
+ * Function to create modals with bootstrap styles for upload avatars.
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt:261]:
```suggestion
 * Creates modals with bootstrap styles for upload avatars.
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/AvatarForm.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+external interface AvatarFormProps : Props {
+    /**
+     * Flag to handle avatar Window
+     */
+    var isOpen: Boolean
+
+    /**
+     * Callback to update state for close window.
+     */
+    var onCloseWindow: (Boolean) -> Unit
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/AvatarForm.kt:33]: do we really need a boolean as argument here?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt [hunk 5]
```diff
@@ -481,18 +479,16 @@ class OrganizationView : AbstractView<OrganizationProps, OrganizationViewState>(
 
     private fun postImageUpload(file: File) =
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt:480]: btw, `postImageUpload` also looks similar for organization and user (only type is different)

**Review by nulls — APPROVED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt [hunk 6]
```diff
@@ -271,3 +256,69 @@ fun ChildrenBuilder.modalBuilder(
...
+                    className = ClassName("modal-body")
+                }
+            }
+        }
+    }
+}
+
+/**
+ * @param title
+ * @param onCloseButtonPressed
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt:299]: Please document the use-case (why and when) `onCloseButtonPressed` may be `null`.

Why is it optional?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt [hunk 7]
```diff
@@ -474,22 +477,19 @@ class OrganizationView : AbstractView<OrganizationProps, OrganizationViewState>(
             it.decodeFromJsonString()
         }
 
-    private fun postImageUpload(element: HTMLInputElement) =
+    private fun postImageUpload(file: File) =
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt:480]: These two versions of `postImageUpload()` (for users and organizations) seem almost identical.

Can we refactor them into a single function which would accept `AvatarType`?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/AvatarForm.kt [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+     * Callback to upload avatar.
+     */
+    var imageUpload: (File) -> Unit
+}
+
+private fun avatarForm() = FC<AvatarFormProps> { props ->
+
+    modalAvatarBuilder(
+        isOpen = props.isOpen,
+        title = "Change organization's avatar",
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/AvatarForm.kt:45]: This is a duplicate string literal (the other occurrence is in the `OrganizationView`).
Can we extract it into a common constant?

Also, organization-wise, _logo_ seems more appropriate than _avatar_.
Up to you, though.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/externals/imageeditor/ReactAvatarImageCropper.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ * External declaration of [reactAvatarImageCropper] react component
+ */
+@JsName("default")
+external val reactAvatarImageCropper: FC<ReactAvatarImageCropperProps>
+
+/**
+ * Props of [ReactAvatarImageCropperProps]
+ */
+@Suppress("TYPE_ALIAS")
+external interface ReactAvatarImageCropperProps : PropsWithChildren {
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/externals/imageeditor/ReactAvatarImageCropper.kt:21]: Why `PropsWithChildren` and not just `Props`?

On the call site, I don't see `reactAvatarImageCropper` having any children.

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt (hunk 5 above)
Comment (0x6675636b796f75676974687562) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt:480]: Indeed.

https://github.com/saveourtool/save-cloud/pull/1711#discussion_r1067947289

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt]: Changed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/AvatarForm.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/AvatarForm.kt]: Removed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt (hunk 5 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt (hunk 7 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/OrganizationView.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/AvatarForm.kt (hunk 8 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/AvatarForm.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/externals/imageeditor/ReactAvatarImageCropper.kt (hunk 9 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/externals/imageeditor/ReactAvatarImageCropper.kt]: Fixed

File: build.gradle.kts [hunk 10]
```diff
@@ -59,7 +59,7 @@ tasks.withType<org.liquibase.gradle.LiquibaseTask>().configureEach {
     @Suppress("MAGIC_NUMBER")
     this.javaLauncher.set(project.extensions.getByType<JavaToolchainService>().launcherFor {
         // liquibase-core 4.7.0 and liquibase-gradle 2.1.1 fails on Java >= 13 on Windows; works on Mac
-        languageVersion.set(JavaLanguageVersion.of(11))
+        languageVersion.set(JavaLanguageVersion.of(13))
```
Comment (kgevorkyan) [inline build.gradle.kts:62]: what about message above? and if all is fine, why `13`? and not `17`, at least?

Comment (nulls) [reply build.gradle.kts]: it fails on Windows, I guess @Cheshiriks uses `Mac`

Comment (Cheshiriks) [reply build.gradle.kts]: I accidentally committed this. Rolled back.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/http/Requests.kt [hunk 11]
```diff
@@ -91,6 +95,29 @@ suspend fun ComponentWithScope<*, *>.getUser(name: String) = get(
...
     .decodeFromJsonString<UserInfo>()
 
+/**
+ * @param file image file
+ * @param name avatar owner name
+ * @param type avatar type
+ */
+suspend fun ComponentWithScope<*, *>.postImageUpload(
+    file: File,
+    name: String?,
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/http/Requests.kt:105]: why name can be `null`?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt (hunk 6 above)
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/modal/ModalBuilder.kt:299]: `onCloseFun` still has `onCloseButtonPressed` as nullable

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/http/Requests.kt (hunk 11 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/http/Requests.kt]: Fixed

**Pull request merged**